### ISSUE TITLE
Improve Tilemap Performance -> allow set_cell via Array

### ISSUE THREAD

**Issue by zzz-assault** [godotengine/godot#31020]
**Godot version:**

- 3.1 (2D)

**OS/device including version:**

- Windows 10
- RTX 2080
- I7 8700

**Issue description:**
For huge procedural Tilemaps the most critical bottle neck is "set_cells", as you need to go via code with a double loop through each cell and call "set_cell" for them.

Unfortunately also multi threading via "Thread" does not work, as when the Tilemap is active in the scene tree it leads to an immediately crash (_active scene tree is NOT thread safe_). The only possitiblity is using a seperate Thread for filling the Tilemap while it's instanced but not active in the scene tree (add_child not executed). But it's not possible to have multiple threads accessing the Tilemap in parallel (Thread Safety) -> will also lead to crash.

I'm aware, that a lot of devs solve this issue, by loading junks of the Tilemap (i.e. while the Player moves), but for RTS and eco simulations you need the full map available at the beginning.

**Solution:**
Create a Tilemap method which allows to set the whole Tilemap cells via loading an Array. If possible for flexibility reasons I would like to get the following options.

- set_cells (Array2D) 
-> should set cell ID accordingly to the Array2D for the full Tilemap
- set_cells(y, Array)
-> should set cell ID accordingly to the Array for the specified line (allows line by line update)
 
Both solutions would in addition to batch update of Tilemap cells also allow multi threading, as Arrays in GDScript can be accessed via multiple threads.

_If there is any other existing solution to create a Tilemap of i. e. 10kx10k in one step with not minutes of loading required, I would be happy to know. I myself wasn't able to come up with an better idea as described above (besides switching to native C++ -> but also C++ would benefit from multi thread access to Tilemap )_

**Minimal reproduction project:**
To reproduce the performance issue just create a new scene, add a Tilemap with a Tileset (only containing one item) and then perform in code with a double loop the loading of i.e. 10.000x10.000 Tiles.

**Comment by Skaruts** [comment]
Could this issue be extended to draw functions like `draw_texture_rect_region` and such, as well as to some form of Sprite batching, where one could set the properties of many sprites in bulk?

**Comment by dioptryk** [comment]
This may be a bit ugly, but can't you generate a _text scene file in memory_, filling it with data you want (even with multiple threads, since the tilemap is just a big int array)? Then you save it raw to "usr://" and load(). Should be orders of magnitude faster than working on a Scene object. 
Of course, this means you become dependent on the serialized TileMap format not changing.

**Comment by Zylann** [comment]
You could theoretically assign the `tile_data` property using `set("tile_data", array)`, where the array is something like this:
```
PoolIntArray( 131077, 0, 0, 131078, 0, 0, 131080, 0, 0, 131081, 0, 0, 131082, 0, 0, 131083, 0, 0, 196613, 0, 0, 196614, 0, 0, 196615, 0, 0, 196616, 0, 0, 262149, 0, 0, 262152, 0, 0, 327685, 0, 0, 327686, 0, 0, 327687, 0, 0, 327688, 0, 0 )
```
That's how tilemaps load from a file, except you would be providing that data by script instead.

**Comment by zzz-assault** [comment]
@Skaruts: I don't see the connection, but be welcome if thus might be combinable ... :)

@dioptryk: Thats a strange but interesting workaround ... should work as long as there is no format difference between the hardcoded and Tilemap text. But as it's anyhow an Array my feature request should be rather simple to implement ... I still have time to wait for a realized feature, as I work on Flame of ZaZuaZ at the moment and this feature would be needed for the next game.

**Comment by zzz-assault** [comment]
@Zylann: That sounds like an approach I could use ... I like the Tilemap implementation and want to use it in my game ... and eventhough the usecase might not be common, the implementation of such a method or documenting your idea (if it's working) should be an easy way to expand the Tilemap usability.

Don't get me wrong, but the engine is promoting GD-Script and also has Thread features, so simply suggesting a language switch to solve performance issues seems to be an odd approach or? (If I would take your statement to the extreme I would question why Thread in GD-Script exists at all) .. 😉

**Comment by Zylann** [comment]
Switching language is sometimes the thing to do. GDScript can work in a very wide range of cases, sometimes the GPU can even be used. But if something is really slow, implementing it in C# or even C++ can outweight the cons.

**Comment by dioptryk** [comment]
@zzz-assault Check out the _tile_map.cpp_, __set_tile_data_ method. If I see correctly, each tile is represented by 3 unsigned integers in the array. First int represents the coordinates (ushort x and ushort y bitshifted into int). The first 29 bits of the second integer represent the tile id in the tileset, the other 3 bits mean flip and transposition. Dunno about the third integer. Since the coordinates don't depend on the position in the array but are decoded, you write the cells in whatever order.

**Comment by zzz-assault** [comment]
@Zylann @dioptryk Thank you both for your feedback so far, at least my tests showed that using set("tile_data", array) would work for multi threading, but unfortunately the algorithm of getting the correct values (interleave with bitshift) is above my knowledge and current understanding of programming … so I'm not able to transfer it into GD-Script.

Anyhow for myself and probably also others trying to utilize Tilemap with the need of loading a huge amount of Tiles at once, I hope that someone might be willing to add the mentioned methods to the Godot Engine.

**Comment by Skaruts** [comment]
@zzz-assault 
> I don't see the connection, but be welcome if this might be combinable ... :)

Well, it's the three places where batching would be useful: the TileMap (for most games), the draw functions (for something like Game Of Life or whatever else) and sprites (for text based roguelikes, for example). In the cases where the TileMap's isn't the way to go you'll be faced with the same problem of having to iterate and set all their properties individually. 

> Another way is to not use Godot's tilemap. It's just one implementation after all, nothing stops you from drawing the tiles yourself :p
The performance is hundreds of frames lower if you do it in gdscript, though. And also, for big maps, too many sprites will inflate map loading times. In my tests a 100x100 roguelike map (10k Sprites) was taking quite a bit to load (which made me switch to emulating a terminal with a fixed amount of sprites: 80x50). 

As for all the other solutions, they are potentially nice workarounds for the meanwhile, but is there any reason why Godot shouldn't support batching? To my experience batching is a common feature in engines and frameworks.

**Comment by zzz-assault** [comment]
@Skaruts understood … but I assume the solution / implementation would be very different, at least for the Tilemap issue (which is anyhow an Array) I assume loading via an Array could be small Change (if I understood the source code correctly).

As said, I have no problem with combining the enhancement requests, if there is any synergy possible … ;)

**Comment by Zylann** [comment]
@Skaruts 

> The raw performance is hundreds of frames lower if you do it in gdscript with Sprites, though

I never said to use sprites :p could be done in `_draw()`, with `draw_texture` or `draw_primitives` with a 2D mesh, and you could use C# is GDScript is the bottleneck.

**Comment by Ranoller** [comment]
I like this idea, it doesn't have to be very difficult to implement. It would only be to move the for loop from gdscript to tile_map.cpp and make the corresponding setters and getters. And the difference in performance would be amazing. Making in the past a random map generator I also concluded that the bottleneck was to call set_cell thousands of times, not in gdscript.
I will look at it (Not promises). Good idea!

**Comment by Skaruts** [comment]
@Zylann fair enough. But you still lose some performance. The TileMap really is quite a boost on tile rendering over the alternatives.

**Comment by zzz-assault** [comment]
@Ranoller Sounds great, perhaps it would also be possible to just load directly the internal Tilemap Array with the input Array instead of executing the loop via set_cells inside the .cpp, but I'm not sure if there would be any dependencies. 

In any case … big THX for having a look into the possible implementation.

@Skaruts I'm not sure how the Godot Engine GitHub issue tracking or organization is done, but as the implementation would be totally different, I would recommend to open a new issue with your request (perhaps even splitting the your two requests, if they also need different implementations?).

**Comment by zzz-assault** [comment]
MrDudeIII on Twitter came up with a correct "interleave with bitshift" and he agreed to share his solution ... so if anyone needs a workaround I'll post it here … :)

![example](https://user-images.githubusercontent.com/47723585/64281849-57dfd480-cf54-11e9-9bb8-078e195a5b1d.png)

**Comment by Sungray** [comment]
This is huge, it makes it possible to load a tilemap in the background using threads and local arrays, stitching the arrays back together into one huge array and dump it into the tilemap instantly with not a single stutter from the main thread. It's a bit of a hack but works perfectly. Should be incorporated into TileMap's methods. Thanks.

**Comment by zzz-assault** [comment]
@Ranoller Any update? :)

**Comment by Ranoller** [comment]
Not working on that, sorry

**Comment by MrDudeIII** [comment]
I feel like should mention that even writing directly to the tile_data wasn't fast enough for my use case (I was trying to update 1000x1000 tiles every frame), so I ended up writing another shader draw each pixel of a 1000x1000 png as a corresponding tile in a 16000x16000 png based on color. This was fast enough (no lag with per-frame updates), but really limited as far as what the tiles can do.

**Comment by Skaruts** [comment]
@zzz-assault 
There's some improvements being made by @lawnjelly at #35957 (and also #35696) that I think you might find interesting.

**Comment by byfron** [comment]
In case it helps anybody. The interleaving format for each cell is: 

[tile coordinates (16 bits each), tile_id (32 bits encoding id(24), flip_x(1), flip_y(1), transpose(1)), and 32 bits with the atlas sub-tile coordinate (16 bits each)]

The details can be checked here: https://github.com/godotengine/godot/blob/master/scene/2d/tile_map.cpp#L1210

**Comment by zzz-assault** [comment]
@Skaruts Thx … but unfortunately it's only valid for GLES2 and it would only solve draw performance (haven't seen a big bottleneck there), biggest problem is still to use a double loop for cell wise loading :(

**Comment by Calinou** [comment]
Closing in favor of https://github.com/godotengine/godot-proposals/issues/1538, as feature proposals are now tracked in the Godot proposals repository.

**Comment by sxdxs** [comment]
So....is it possible to do like oblique tiles? Pulling my hair out trying on all the engines

**Comment by MrDudeIII** [comment]
@sxdxs I'm not sure what you mean by opaque. Also, you might want to try asking that question on the official Godot engine discord. I'm not sure you'll get a timely response here, and your question seems pretty off topic from this issue.
